--- node_modules/baguettebox.js/dist/baguetteBox.js
+++ node_modules/baguettebox.js/dist/baguetteBox.js
@@ -194,9 +194,9 @@
                 tagsNodeList = galleryElement.getElementsByTagName('a');
             }

-            // Filter 'a' elements from those not linking to images
+            // Filter 'a' elements from those not linking to images, and not containing <img> tags (PATCHED)
             tagsNodeList = [].filter.call(tagsNodeList, function(element) {
-                if (element.className.indexOf(userOptions && userOptions.ignoreClass) === -1) {
+                if (element.className.indexOf(userOptions && userOptions.ignoreClass) === -1 && element.getElementsByTagName("img").length > 0) {
                     return regex.test(element.href);
                 }
             });
